ResponseFuture: do not return the stream ID on client timeout

piodul opened the pull request:
When a timeout occurs, the ResponseFuture associated with the query
returns its stream ID to the associated connection's free stream ID pool,
so that the stream ID can be immediately reused by another query.

However, that is incorrect and dangerous. If query A times out before it
receives a response from the cluster, a different query B might be
issued on the same connection and stream. If response for query A
arrives earlier than the response for query B, the first one might be
misinterpreted as the response for query B.

This commit changes the logic so that stream IDs are not returned on
timeout - now, they are only returned after receiving a response.

psarna commented:
I tried the updated driver on the reproducer. What happens right now is that the third request, which previously incorrectly reused a stream id, now fails with `'Unable to complete the operation against any hosts', {<Host: 127.0.0.1:9042 datacenter1>: AssertionError()}`. Is it the desired outcome? It's infinitely better than incorrect results, but perhaps it should be queued and run once there are enough stream ids available?

psarna commented:
edit: never mind! I tested based on 3.24, without this fix present:
```
commit a6a66cac2931c4f66d248ed7091d68f6932de5b0
Author: Alan Boudreault <alan@kovaro.ca>
Date:   Tue Jun 2 09:46:04 2020 -0400

    ensure the connection max request id's is respected

diff --git a/cassandra/connection.py b/cassandra/connection.py
index f30be682..3d154de0 100644
--- a/cassandra/connection.py
+++ b/cassandra/connection.py
@@ -1443,7 +1443,7 @@ class HeartbeatFuture(object):
         log.debug("Sending options message heartbeat on idle connection (%s) %s",
                   id(connection), connection.endpoint)
         with connection.lock:
-            if connection.in_flight <= connection.max_request_id:
+            if connection.in_flight < connection.max_request_id:
                 connection.in_flight += 1
                 connection.send_msg(OptionsMessage(), connection.get_request_id(), self._options_callback)
             else:
diff --git a/cassandra/pool.py b/cassandra/pool.py
index 87e8f037..cd276560 100644
--- a/cassandra/pool.py
+++ b/cassandra/pool.py
@@ -418,7 +418,7 @@ class HostConnection(object):
         remaining = timeout
         while True:
             with conn.lock:
-                if conn.in_flight <= conn.max_request_id:
+                if conn.in_flight < conn.max_request_id:
                     conn.in_flight += 1
                     return conn, conn.get_request_id()
             if timeout is not None:
diff --git a/tests/integration/simulacron/test_connection.py b/tests/integration/simulacron/test_connection.py
index afe2685d..11bfef7f 100644
--- a/tests/integration/simulacron/test_connection.py
+++ b/tests/integration/simulacron/test_connection.py
@@ -24,7 +24,7 @@ from mock import Mock, patch
 from cassandra import OperationTimedOut
 from cassandra.cluster import (EXEC_PROFILE_DEFAULT, Cluster, ExecutionProfile,
                                _Scheduler, NoHostAvailable)
-from cassandra.policies import HostStateListener, RoundRobinPolicy
+from cassandra.policies import HostStateListener, RoundRobinPolicy, WhiteListRoundRobinPolicy
```

... and then, an assertion is hit when trying to get another stream_id due to the off-by-one error. With the patch above applied, the error no longer happens.

An important conclusion is that if we backport the fix presented in this issue to previous versions, we should also ship the patch I posted above.

psarna commented:
And just to clarify - after both fixes are in place, the code works perfectly - in case of max concurrency == 2, two requests are performed first, and only then, the third one is sent - even though one of the first two timed out.

piodul commented:
I think that the PR might still have some subtle synchronization issues. I'll look through the code once more - please do not merge yet.

haaawk reviewed (changes requested):
LGTM in general but we need a way to close the connection and open another one when we accumulate two many orphaned_stream_ids. Just like a Java driver does:
https://github.com/datastax/java-driver/commit/4e680fbde94366b9a213f12253cfabf7c4aea239

```
Motivation:

Sometimes the driver might stop waiting for a response on a particular
channel (for example if the request timed out, or was completed by
another speculative execution). Before this change we just kept the
response callback in our inflight map, which is a problem if we never
get the response: it creates a memory leak, and if this keeps happening
we'll eventually run out of stream ids on the channel.

Modifications:

Add a channel method to cancel a response callback, indicating that the
caller is not interested in the response anymore. Modify existing
clients to cancel their callbacks.
Track the number of "orphan" stream ids (cancelled and have not yet
received a response from the server). Initiate a graceful shutdown if
this number exceeds a threshold.
Improve the way the channel pool manages channel shutdowns: start the
reconnection as soon as an orderly shutdown has *started*, not when it
finishes.

Result:

Cancelled callbacks are not leaked anymore. When the number of orphan
ids exceed the threshold, the channel is closed gracefully and the pool
starts replacing it immediately.
```

piodul commented:
v2: Added commits with fixes to two issues I found by reading the code more thoroughly. One caused the connection to be re-created when we run out of free streams on it and there are connections timed out on the client side but not replied by Scylla. The other caused more timeouts than necessary right after the free stream pool gets empty. More information about them in the commit descriptions.

piodul commented:
It looks like CI checks are skipped for now... For this PR, I ran the following tests locally:

- Unit tests: `nosetests -w tests/unit/`
- Integration tests specified in the `.github/workflows/integration-tests.yaml` file
- Some manual unit tests which hits the per-connection request limit and runs enough requests concurrently so that they time out

I can run more if requested, although it would be nice if the tests were ran in a standardized environment...

avelanarius commented:
> I can run more if requested, although it would be nice if the tests were ran in a standardized environment...

It seems you need to add "integration-tests" tag (which I just did) to run the tests - however it still didn't help...

piodul commented:
v3:
- Implemented excess connection pooling (improves robustness of connecting to the right shards, it is also needed for the connection replacement mechanism to work well),
- Implemented connection replacement similar to the one mentioned [here](https://github.com/scylladb/python-driver/pull/106#pullrequestreview-765181702) - when the number of orphaned stream IDs crosses a certain threshold, a new connection is opened which replaces the old one, and the old connection is closed after all its outstanding requests either finish or time out.

psarna commented:
@piodul here's a neat trick for pinging: @scylladb/python-driver-maint
